### PR TITLE
APEXMALHAR-2130 WindowStorage interface changes in preparation of inc…

### PR DESCRIPTION
**Pull request by davidyan74**
…orporating spillable data structures

@siyuanh please review and merge
@tweise @ShunxinLu FYI

### CODE REVIEW

File: library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java [hunk 1]
```diff
@@ -45,11 +45,12 @@
    */
   class PlainTuple<T> implements Tuple<T>
   {
-    private T value;
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java:48]: Why change this to final? What if I have a stateless transform in between 2 windowed operator and I want to change the value of the tuple. Then I have to recreate the Tuple instance

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java [hunk 2]
```diff
@@ -148,20 +136,49 @@ public boolean equals(Object other)
...
+      } else if (this.getDurationMillis() > o.getDurationMillis()) {
+        return 1;
+      }
+      return 0;
+    }
+
+    @Override
+    public String toString()
+    {
+      return "[TimeWindow " + getBeginTimestamp() + "(" + getDurationMillis() + ")]";
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:164]: It is better to be in json formatter

File: library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java [hunk 3]
```diff
@@ -53,41 +50,87 @@
...
+     * @return
+     */
+    T get(Window window);
+
+    /**
+     * Returns the iterable of the entries in the storage
+     *
+     * @return
+     */
+    Iterable<Map.Entry<Window, T>> entries();
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java:89]: This method always return itself?

Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java:89]: Actually I'd rather change the method to allWindows and remove Iterable from class signature. It's hard to understand a WindowedPlainStorage is iterable.

File: library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java [hunk 4]
```diff
@@ -53,41 +50,87 @@
...
+     */
+    void put(Window window, K key, V value);
+
+    /**
+     * Gets the iterator over the key/value pairs associated with the given window
+     *
+     * @param window
+     * @return
+     */
+    Iterator<Map.Entry<K, V>> iterator(Window window);
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java]: what's the difference between iterator and entries?

File: library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java (hunk 3 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java:89]: @siyuanh  makes sense. It's updated.

File: library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java (hunk 4 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java]: @siyuanh removed iterator

File: library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java (hunk 1 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java:48]: @siyuanh you have a point. @tweise @vrozov do you guys know of any value in making tuples immutable?

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java (hunk 2 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:164]: I agree but that's not what the convention is in the rest of apex code. This is more of a convenience for logging so it's better to make it a single line string, even though we can make json one line as well.

File: library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java (hunk 1 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java:48]: @siyuanh  I have reverted this change for now.

**Comment by davidyan74** [comment]
@siyuanh please take a look at this again and see if you have any more issues before this can be merged

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java [hunk 5]
```diff
@@ -148,20 +136,49 @@ public boolean equals(Object other)
...
+    }
   }
 
   /**
    * SessionWindow is a window that represents a time slice for a key, with the time slice being variable length.
    *
    * @param <K>
    */
-  class SessionWindow<K> extends TimeWindow
+  class SessionWindow<K> extends TimeWindow<SessionWindow<K>>
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:173]: javadoc for compareTo here

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java [hunk 6]
```diff
@@ -59,54 +68,33 @@ public boolean equals(Object other)
...
-  /**
-   * The singleton default comparator of windows
-   */
-  Comparator<Window> DEFAULT_COMPARATOR = new DefaultComparator();
-
-  /**
    * TimeWindow is a window that represents a time slice
    */
-  class TimeWindow implements Window
+  class TimeWindow<WINDOW extends TimeWindow> implements Window<WINDOW>
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/Window.java]: javadoc for compareTo

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java (hunk 2 above)
Comment (siyuanh) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:164]: override in SlidingTimedWindow

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java [hunk 7]
```diff
@@ -389,8 +397,32 @@ public void processWatermark(ControlTuple.Watermark watermark)
   @Override
   public void setup(Context.OperatorContext context)
   {
-    this.windowWidthMillis = context.getValue(Context.DAGContext.STREAMING_WINDOW_SIZE_MILLIS);
+    this.timeIncrement = context.getValue(Context.OperatorContext.APPLICATION_WINDOW_COUNT) * context.getValue(Context.DAGContext.STREAMING_WINDOW_SIZE_MILLIS);
     validate();
+    windowStateMap.setup(context);
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:402]: I'm thinking if we could provide an automatic way to call these method by set annotation on the field just like those IoC frameworks do

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/InMemorySessionWindowedStorage.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  }
+
+  @Override
+  public Collection<Map.Entry<Window.SessionWindow<K>, V>> getSessionEntries(K key, long timestamp, long gap)
+  {
+    List<Map.Entry<Window.SessionWindow<K>, V>> results = new ArrayList<>();
+    TreeSet<Window.SessionWindow<K>> sessionWindows = keyToWindows.get(key);
+    if (sessionWindows != null) {
+      for (Window.SessionWindow<K> window : sessionWindows) {
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/impl/InMemorySessionWindowedStorage.java]: This might be resolved in O(logn) time

File: library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java (hunk 1 above)
Comment (vrozov) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Tuple.java:48]: It depends on how PlainTuple is used later. If it is used as a key, making PlainTuple immutable may be helpful. But I guess, T itself may be used as a key, instead of PlainTuple. In cases mutable or immutable use case becomes common, we may introduce (Im)MutablePlainTuple.

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java (hunk 7 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:402]: not the scope of this PR :)

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java (hunk 5 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:173]: @siyuanh done

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/InMemorySessionWindowedStorage.java (hunk 8 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/InMemorySessionWindowedStorage.java]: @siyuanh fixed

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java (hunk 2 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:164]: WindowOption.SlidingTimeWindows generates Window.TimeWindow itself. There is no class that represents individual sliding time windows.

**Comment by siyuanh** [comment]
@davidyan74 @tweise @ShunxinLu I have no more questions about the pull request. If you guys are ok with the change. I will go ahead and merge it.

File: demos/highlevelapi/src/main/java/org/apache/apex/malhar/stream/sample/complete/TwitterAutoComplete.java [hunk 9]
```diff
@@ -125,7 +125,7 @@ public ComputeTopFlat(int candidatesPerPrefix, int minPrefix)
           public Tuple<KeyValPair<String, CompletionCandidate>> f(KeyValPair<String, CompletionCandidate> tuple)
           {
             // TODO: Should be removed after Auto-wrapping is supported.
-            return new Tuple.WindowedTuple<>(Window.GLOBAL_WINDOW, tuple);
+            return new Tuple.WindowedTuple<>(Window.GlobalWindow.getInstance(), tuple);
```
Comment (tweise) [inline demos/highlevelapi/src/main/java/org/apache/apex/malhar/stream/sample/complete/TwitterAutoComplete.java]: Why this change?

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java [hunk 10]
```diff
@@ -18,31 +18,42 @@
...
+    /**
+     * The singleton global window
+     */
+    private static GlobalWindow GLOBAL_WINDOW = new GlobalWindow();
+
     private GlobalWindow()
     {
     }
 
+    public static GlobalWindow getInstance()
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/window/Window.java]: This is not needed, rename GLOBAL_WINDOW to INSTANCE and make it public.

Comment (siyuanh) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Window.java]: I'm also curious about this change and to me. And also the equals method can use == instead of instanceof which I believe would be slightly faster

**Comment by davidyan74** [comment]
all comments addressed. @siyuanh  please merge if you see no other issues.

File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java [hunk 11]
```diff
@@ -18,27 +18,33 @@
...
+   * Global window means there is only one window that spans the entire life time of the application
    */
-  class GlobalWindow implements Window
+  class GlobalWindow implements Window<GlobalWindow>
   {
+
+    /**
+     * The singleton global window
+     */
+    public static GlobalWindow INSTANCE = new GlobalWindow();
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/window/Window.java]: Should be final

Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/Window.java]: @siyuanh added

**Comment by tweise** [comment]
@davidyan74 LGTM please rebase.
